`image:push` to  private registries is broken when using hostname/repo

I believe you should be able to push to a private registry by using the full name of the image that includes the host/port/path to the registry.  

`dobi` supports the same auth config as the docker client, so authentication should work as well.

I need to test it out to confirm.

cc @jmmills (let's move the discussion here, out of https://github.com/docker/compose/issues/1126)

@dnephin Tried that, blows up as invalid repository/tag:
```
[DEBUG] [image:build cattle-prod] http://my.private.registry.net/cattle-prod time=2016-11-02 19:00:17.568532586 -0400 EDT Start
[DEBUG] [image:build cattle-prod] http://my.private.registry.net/cattle-prod Image does not exist
[DEBUG] [image:build cattle-prod] http://my.private.registry.net/cattle-prod is stale
[DEBUG] stopping tasks
[ERROR] Failed to execute task "cattle-prod:build": API error (500): Error parsing reference: "http://my.private.registry.net/cattle-prod:42475a955b" is not a valid repository/tag
```
In our case, yes, we use the convention: my.private.registry.net/{repository}/{service}:{short-sha}

Our current process involves doing `docker build -t my.private.registry.net/{repository}/{service{:{short-sha} ./directory` and `docker push my.private.registry.net/{repository}/{service}:{short-sha}` which works perfectly fine using the docker client locally and on our build server.

"cattle-prod" is done a bit differently (probably mistakenly), as you can see, with just the repository name followed by a version tag. I just tried this with: `image: my.private.registry.net/cattle-prod/cattle-prod` and it did work perfectly. The logic holds up for the "correct" use of slashes by any normal standard on our part, too. I think we just hit and edge case here.

I do want to plus the variable interpolation in the `image` field though, that could be very useful for DRYing some build/test jobs and finding commonality across different repositories.

--- a/tasks/image/utils.go
+++ b/tasks/image/utils.go
@@ -33,16 +33,17 @@ func GetCanonicalTag(ctx *context.ExecuteContext, conf *config.ImageConfig) stri
 }
 
 func parseAuthRepo(image string) (string, error) {
-	// This is the approximate logic from
-	// github.com/docker/docker/reference.splitHostname(). That package is
-	// conflicting with other dependencies, so it can't be imported at this time.
-	parts := strings.SplitN(image, "/", 3)
-	switch len(parts) {
-	case 1, 2:
+	return splitHostname(image), nil
+}
+
+// Copied from github.com/docker/docker/reference/reference.go
+// That package is conflicting with other dependencies, so it can't be imported
+// at this time.
+func splitHostname(name string) string {
+	i := strings.IndexRune(name, '/')
+	if i == -1 || (!strings.ContainsAny(name[:i], ".:") && name[:i] != "localhost") {
 		logging.Log.Debugf("Using default registry %q", defaultRepo)
-		return defaultRepo, nil
-	default:
-		logging.Log.Debugf("Using registry %q", parts[0])
-		return parts[0], nil
+		return defaultRepo
 	}
+	return name[:i]
 }

--- a/tasks/image/utils_test.go
+++ b/tasks/image/utils_test.go
@@ -23,3 +23,9 @@ func TestParseAuthRepoPrivateRepo(t *testing.T) {
 	assert.Nil(t, err)
 	assert.Equal(t, repo, "myrepo.net")
 }
+
+func TestParseAuthRepoPrivateRepoNoUsername(t *testing.T) {
+	repo, err := parseAuthRepo("myrepo.net/foo")
+	assert.Nil(t, err)
+	assert.Equal(t, repo, "myrepo.net")
+}

